pkgs/test_api/lib/hooks_testing.dart

```diff
@@ -0,0 +1,156 @@
+  final _done = Completer<void>();
+  TestCaseMonitor._(FutureOr<void> Function() body)
+      : _liveTest = _createTest(body);
+
+  /// Run [body] as a test case and return a [TestCaseMonitor] with the result.
+  ///
+  /// The [state] will either [State.passed], [State.skipped], or
+  /// [State.failed], the test will no longer be running.
+  ///
+  /// {@template result-late-fail}
+  /// Note that a test can change state from [State.passed] to [State.failed]
+  /// if the test surfaces an unawaited asynchronous error.
+  /// {@endtemplate}
+  ///
+  /// ```dart
+  /// final monitor = await TestCaseMonitor.run(() {
+  ///   fail('oh no!');
+  /// });
+  /// assert(monitor.state == State.failed);
+  /// assert((monitor.errors.single.error as TestFailure).message == 'oh no!');
+  /// ```
+  static Future<TestCaseMonitor> run(FutureOr<void> Function() body) async {
+    final monitor = TestCaseMonitor.start(body);
+    await monitor.onDone;
+    return monitor;
+  }
+
+  /// Start [body] as a test case and return a [TestCaseMonitor] with the status
+  /// and result.
+  ///
+  /// The [state] will start as [State.pending] if queried synchronously, but it
+  /// will switch to [State.running]. After `onDone` completes the state will be
+  /// one of [State.passed], [State.skipped], or [State.failed].
+  ///
+  /// {@macro result-late-fail}
+  ///
+  /// ```dart
+  /// late void Function() completeWork;
+  /// final monitor = TestCaseMonitor.start(() {
+  ///   final outstandingWork = TestHandle.current.markPending();
+  ///   completeWork = outstandingWork.complete;
+  /// });
+  /// await pumpEventQueue();
+  /// assert(monitor.state == State.running);
+  /// completeWork();
+  /// await monitor.onDone;
+  /// assert(monitor.state == State.passed);
+  /// ```
+  static TestCaseMonitor start(FutureOr<void> Function() body) =>
+      TestCaseMonitor._(body).._start();
+
+  void _start() {
+    _liveTest.run().whenComplete(_done.complete);
+  }
+
+  /// A future that completes after this test has finished running, or has
+  /// surfaced an error.
+  Future<void> get onDone => _done.future;
+
+  /// The running and success or failure status for the test case.
+  State get state {
+    final status = _liveTest.state.status;
+    if (status == Status.pending) return State.pending;
+    if (status == Status.running) return State.running;
+    final result = _liveTest.state.result;
+    if (result == Result.skipped) return State.skipped;
+    if (result == Result.success) return State.passed;
+    // result == Result.failure || result == Result.error
+    return State.failed;
+  }
+
+  /// The errors surfaced by the test.
+  ///
+  /// A test with any errors will have a failing [result].
+  ///
+  /// {@macro result-late-fail}
+  ///
+  /// A test may have more than one error if there were unhandled asynchronous
+  /// errors surfaced after the test is done.
+  Iterable<Object> get errors => _liveTest.errors.map((a) => a.error);
```

Comment:
@jakemac53 @grouma - the only current usage for looking at the `StackTrace` from the errors is to forward them so they are printed for unexpected error.

https://github.com/dart-lang/test/blob/ba8e43afa183e321a482efdea06e80d07d06c0dc/pkgs/test_api/test/utils.dart#L115

I don't know if that is worthwhile enough to add an indirection for the vast majority of uses cases that don't care about the stack trace. If our own tests in `matcher` or `checks` are failing with unexpected exception the stack trace could end up being helpful though. In the end I landed on not exposing the stack trace, so that we don't need to worry about returning a class from `package:async`.

WDYT?

Reply:
> If our own tests in matcher or checks are failing with unexpected exception the stack trace could end up being helpful though

How so?

Reply:
It doesn't make any difference for passing tests. If we have some bug in a test or in library code that causes an unexpected error within the `liveTest` body you'd see a difference in the output.  Currently we'd include a stack trace for the error, and it would show up like any other test failure pointing directly to the cause.

If we strip the `StackTrace` from this view, then when we forward the errors we won't see them.

Reply:
That seems valuable imo and worth keeping the `StackTrace` around.

Reply:
Yeah I think it would be beneficial to keep around the stack trace. This api shouldn't be used by most people anyways right? So it shouldn't be much of a pain that it is exposing something from package:async.

The biggest risk I see is people may rely on inference to get the type and not have a dependency on async themselves, exposing their package to potential breakage, but that is something I can live with (and is really just a general issue).

Reply:
Fair enough. I restored the stack trace. I have found myself adding more stack trace outputs in `checks` as I go, so that's a pretty good indication it's worthwhile here too.